packages/task/src/browser/task-configuration-manager.ts

```diff
@@ -204,6 +219,26 @@ export class TaskConfigurationManager {
             return selected.content;
         }
     }
+
+    protected readonly toDisposeOnDelegateChange = new DisposableCollection();
+    protected updateWorkspaceModel(): void {
+        const newDelegate = this.workspaceService.saved ? this.workspacePreferences : this.folderPreferences;
+        const effectiveScope = this.workspaceService.saved ? TaskScope.Workspace : this.workspaceService.tryGetRoots()[0]?.resource.toString();
```

Comment:
@colin-grant-work 
Hello!
Could you share what use case here is handled.

I thought that the use case is `Untittled (Workspace)` - so unsaved:
![untitled](https://user-images.githubusercontent.com/5676062/115567997-872b8c80-a2c4-11eb-9db9-352cfcd9e288.png)

but I get  `newDelegate = this.workspacePreferences` and `effectiveScope = TaskScope.Workspace` for such use case.

What steps I should do to get:
- newDelegate = this.folderPreferences;
- effectiveScope = this.workspaceService.tryGetRoots()[0]?.resource.toString();
 
Thanks in advance!

Reply:
Hey, @RomanNikitenko! As soon as a workspace file exists, the workspace is considered `.saved`, even if you haven't officially named it and saved it yourself. The case in which the delegate is folder preference is in a single-root unsaved workspace - i.e. the case in which the user uses 'Open folder...' to open a single folder as the workspace.

There was a slight hiccup with this logic where the `TaskConfigurationManager` wasn't switching to the workspace preferences delegate as soon as a workspace file was created, but it that was fixed recently in #9331. It looks like you're seeing that change in action: as soon as you have a workspace file, whether you've titled it or not, the `TaskConfigurationManager` starts considering the workspace file to be the source of information for workspace-scoped tasks.

Reply:
thank you for the quick answer!
one more question:
> The case in which the delegate is folder preference is in a single-root unsaved workspace - i.e. the case in which the user uses 'Open folder...' to open a single folder as the workspace.

If user has a single folder - then tasks are considered on the folder level and should be handled [here](https://github.com/eclipse-theia/theia/blob/7d3b1d3b33cb565ec07965e5baf10c926006cdfd/packages/task/src/browser/task-configuration-manager.ts#L124)

So, I still don't see what use case we handle here.
I'm not saying that there is a problem - I'm investigating some area related to multi-root workspace and have found this place - so I just want to clarify it for myself.
thanks!

Reply:
You're right that all folder-scoped tasks, in both saved and unsaved workspaces, are handled by the folder preference provider. In the case of a single-root unsaved workspace, workspace scope *also* points to folder scope for the single folder. This parallels the preference system, which also collapses folder and workspace scope in the case of a single-root unsaved workspace. Since I'm not sure what your use case is, I'm not sure what steps to provide to 'reproduce' any particular behavior, but if you want to see how it works, you can do something like this:

1. Create a command that does something like this:
```typescript
{
    execute: () => {
        console.log("User scope:", this.taskConfigurationManager.getTasks(TaskScope.Global));
        console.log("Workspace scope:", this.taskConfigurationManager.getTasks(TaskScope.Workspace));
        for (const root of this.workspaceService.tryGetRoots()) {
            console.log(`Folder scope (${root.resource.path.base}):`, this.taskConfigurationManager.getTasks(root.resource.toString()));
        }
    }
}
```
Try that command in different kinds of workspaces. In a single-root unsaved workspace, the logs for workspace and folder scope should be the same (same `_scope` fields, too). In a saved workspace, they should be different. There aren't very many services that interface directly with the `TaskConfigurationManager` - mostly its output gets filtered through the `TaskService`, which is able to identify the fact that workspace scope and folder scope are duplicated in a single-root workspace (because the `label` and `_scope` fields are identical), so consumers of the data end up seeing what they should see: one set of folder-scoped tasks.

Conceivably, in a single-root unsaved workspace, we could just return an empty array for workspace tasks, but I prefer to keep the behavior of the various preference-related functionality consistent. Since the preferences service collapses workspace scope and folder scope in single-root workspaces, I think tasks (and `launch`), should do so, as well.

Reply:
thank you for the detailed answer.

I think I was confused for the single-root use case(user have opened just a single folder) because
**before your explanation I thought** that:
- for the single-root use case tasks should be handled [here](https://github.com/eclipse-theia/theia/blob/7d3b1d3b33cb565ec07965e5baf10c926006cdfd/packages/task/src/browser/task-configuration-manager.ts#L124) on the `Folder` level - there is no need to create `TaskConfigurationModel` for the `Workspace` scope. 
- there should be no `Workspace` scoped tasks for the such use case - as they are duplicated - in the example that you provided above `taskConfigurationManager` returns the same tasks for the `Workspace` scope and for the `Folder` level

Thanks again for your time and efforts - now it's more clear for me!